### ISSUE TITLE
(React) UI Components don't support password managers

### ISSUE THREAD

**Issue by Purii** [aws-amplify/amplify-js#5782]
**Describe the bug**
Browsers don't recognize the input elements as Auth related.

-> No "password suggestion" on during account creation
-> No autofill during sign in
-> No prompt "save password?" during sign in

I'm not sure what details are necessary for this to work, but I tried setting the `name` attribute and others through `formFields`. But that didn't help solve this issue. (https://www.chromium.org/developers/design-documents/form-styles-that-chromium-understands)

-> Form is not part of a "form" element
-> The "sign in/up" buttons aren't labelled as type "submit"

**Comment by Purii** [comment]
Seems that the usage of `shadow DOM` is the root cause: https://bugs.chromium.org/p/chromium/issues/detail?id=770175

**Comment by anantoghosh** [comment]
[Vaadin login component](https://vaadin.com/components/vaadin-login/java-examples)  handles this by not using the shadow dom for the form
https://github.com/vaadin/vaadin-login/blob/9e76baf5e8a7dccf5e65116d0fc91b49ea3bdf2a/src/vaadin-login-form.html#L50

**Comment by PeterBurner** [comment]
does not work with `@aws-amplify/ui-vue` either

**Comment by richiewebgate** [comment]
same issue with aws-amplify-angular

**Comment by kjellski** [comment]
This looks like a bug to every end-user - any hints on how to help fix this here?

**Comment by BrnoPCmaniak** [comment]
So I tried to fix this. Following upon @anantoghosh I googled and found out that the most likely issue is that the form fields are scattered across different shadow DOM roots.

I also noticed this commit https://github.com/aws-amplify/amplify-js/commit/538ca667296568fc58d77ce5c45f201b59e085ea

So I tried to undo some of it works. And instead of using shadow dom for the whole thing I tried to force it to use fallback (`scoped: true` https://stenciljs.com/docs/component#component-decorator).

When I applied this to authenticator and everything underneath this have happend:

![Pokus-1](https://user-images.githubusercontent.com/1249759/94354258-d3d61e80-0079-11eb-91d2-acd8280068f5.png)

```
<amplify-authenticator>
  <amplify-sign-in slot="sign-in" hideSignUp="true" [formFields]="this.signInFormFields"></amplify-sign-in>
</amplify-authenticator>
```

Then I tried to apply it only from form down. Which initialy seemed to be working. Both chrome and firefox now saves the password properly. Unfortunately in chrome you can not use saved credentials, on firefox this works.

I'm out of ideas maybe someone can make it work.

https://github.com/BrnoPCmaniak/amplify-js/commit/fe91c7ea71d1efaa73a8923864a2b01c2e70cc33

**Comment by TAConsulting** [comment]
Any updates on this?

**Comment by simon-lanf** [comment]
Please fix this, it is also affecting almost every testing frameworks. You can't use Selenium, Cypress and stuff like that.

**Comment by adamjmarkham** [comment]
This is a blocker for us (and probably most people) to upgrade our version of Amplify without a workaround. Looks like this bug seems to have been there for a while, any news on when it will be fixed?

**Comment by ericclemmons** [comment]
There's a lot of research in this thread to look into, thanks for that!

In the meantime, @simon-lanf can you provide more info on you how Cypress isn't working?  We have Cypress tests using commands like the following successfully:

```js
cy.get(selectors.usernameInput).type(COGNITO_SIGN_IN_USERNAME);
cy.get(selectors.signInPasswordInput).type(COGNITO_SIGN_IN_PASSWORD);
cy.get(selectors.signInSignInButton).contains('Sign In').click();
``

**Comment by p0fi** [comment]
Any news on this topic? This is a **really** annoying bug!

**Comment by ryan-mars** [comment]
This makes `@aws-amplify/ui-react` a no-go for us. Password managers are a healthy security pattern.

Happy to contribute but I'm not sure where to start.

**Comment by ryan-mars** [comment]
@BrnoPCmaniak does this change work? ---> https://github.com/BrnoPCmaniak/amplify-js/commit/fe91c7ea71d1efaa73a8923864a2b01c2e70cc33

**Comment by BrnoPCmaniak** [comment]
@ryan-mars The issue is using of shadow dom for the form (not using plain HTML for the form). I tried to force the components to be in compatibility mode. 

It works on Firefox. On chrome, it only saves the password but it won't let you use it. 

It could be hacked together with this https://developers.google.com/web/fundamentals/security/credential-management to make it work but it's not a solution.

**Comment by Purii** [comment]
Slots might help fixing the issue: https://stenciljs.com/docs/templating-jsx#slots
That way the Input field is not "hidden" in the shadow dom.

Unfortunately I can't spare the time to try it out at the moment..

**Comment by fbiel** [comment]
Personally, I don't think this should be labeled as a feature-request, this is a bug. Users are really confused about this.

Currently, there is no point in using the password component.

**Comment by DavidNelsonCordelia** [comment]
Any progress on this?  its a unpleasant missing feature.

**Comment by richiewebgate** [comment]
I will unsubscribe now, because this will never be fixed.

**Comment by 24601** [comment]
The use of the shadow DOM is highly problematic, not just for password managers, but for testability, as DOM selectors need to be manually altered to work around the shadow DOM. 

This should no doubt be a bug.

**Comment by wlee221** [comment]
Hello all, we apologize about the lack of response here. As pointed out, this is due to the use of shadow DOM. We will mark this as a bug.

We found that disabling the shadow DOM with _current_ UI components are problematic, because slots (which we use to allow customization) do not work well with shadow DOM disabled. In particular, conditional rendering (https://github.com/ionic-team/stencil/issues/848), default rendering (https://github.com/ionic-team/stencil/issues/2054), and nested slots (https://github.com/ionic-team/stencil/issues/399) do not work, which are patterns we use throughout the library.

For that reason, we are currently working on rewriting our ui-components to not use Web Components / shadow DOM. This will make the form elements accessible to browsers and enable password managers to work. We will hold a public RFC and create a public repo rewrite soon to solicit feedbacks. We will let you know when we publish the RFC.

**Comment by wlee221** [comment]
With regards to our current components (before we publish the rewrite), we are looking at workaround that’ll allow saving username/password and autofilling credentials based on that ([branch](https://github.com/aws-amplify/amplify-js/tree/support-pw-manager)). 

![Kapture 2021-04-20 at 16 46 09 (1)](https://user-images.githubusercontent.com/43682783/116116178-b862de80-a66f-11eb-8a17-e37eac4df024.gif)

This workaround does not support dropdown for multiple credentials, but this will at least allow for a credential save and autofill while we work on the rewrite. Please let us know with any feedback or suggestions. We'll work on getting the workaround tested and reviewed meanwhile.

**Comment by wlee221** [comment]
Hello all, we just merged in the password manager workaround (pr #8181). This is now available in `unstable` npm tag and will be published in the next release cycle. To use this workaround, you need to wrap your authenticator with `AmplifyAuthContainer`:

```tsx
<AmplifyAuthContainer>
  <AmplifyAuthenticator>
    ...
  </AmplifyAuthenticator>
</AmplifyAuthContainer>
``` 
(or with `amplify-auth-container` on vue/angular). Under the hood, `AmplifyAuthContainer` contains a hidden form that is not shadowed and thus gets autofilled. Then we use event listeners to send that autofill down to the amplify-authenticator.

This workaround works for Chrome and Firefox. Safari does not trigger credentials save even when form shadows are disabled, so this will be a limitation until our complete rewrite. Please let us know with any questions or issues with this solution. We'll keep this issue open to track this until this is completely resolved.

**Comment by cburkins** [comment]
I'm currently enjoying the simplicity of using the withAuthenticator HOC.  As such, am I easily able to deploy this ?    I'm not sure how to apply "you need to wrap your authenticator with AmplifyAuthContainer"

**Comment by wlee221** [comment]
@cburkins, you would have to eject out of `withAuthenticator` to use this workaround:

```diff
- const App = () => (<div>My App</div>);
- export default withAuthenticator(App, {initialAuthState: 'signup'});

+ const App = () => {
+   <AmplifyAuthContainer>
+     <AmplifyAuthenticator initialAuthState="signup">
+        <div>My App</div>
+     </AmplifyAuthenticator>
+   </AmplifyAuthContainer>
+ }
```

**Comment by wlee221** [comment]
@cburkins, we've given this more thought and we will be modifying the HOC to include this workaround by default. We'll create a PR soon.

**Comment by wlee221** [comment]
Hi all, I've opened #8243, which will include this workaround to `withAuthenticator` by default. We'll get this reviewed and merged this week.

I've also published the patch to `ui-preview` npm tag for testing purposes. Please note that you can use this tag to unblock yourself and try this out. Let us know if you catch any unexpected behaviors. Thank you!

```
yarn add aws-amplify@ui-preview @aws-amplify/ui-react@ui-preview
```

**Comment by wlee221** [comment]
You're welcome! We just merged #8243 to support `withAuthenticator` -- you shouldn't need to update any code to use this workaround. This has some restrictions as stated before, but hopefully this will alleviate some of the accessibility pain while we work on the rewrite. 

The fix will hit `unstable` soon and will be published in the next release cycle. We'll fix this issue more completely on our rewrite (without hidden forms 😄 ) and close this issue then. I'll mark this back to feature-request as a workaround has been provided. Thanks for your patience!

**Comment by cburkins** [comment]
@wlee221  Thanks !   Tested with native Chrome autofill, and seems to work !   Am I correct in saying that this won't currently work within 3rd party password managers like LastPass ?   Doesn't seem like it works.....

**Comment by wlee221** [comment]
Right, this workaround is intended for browser password managers. 

Support for third party password manager really depends on their implementation and whether they can transverse through the top level shadow DOM or not. For instance 1password supports shadow DOM ([discussion](https://1password.community/discussion/79137/shadow-dom-polymer-forms-do-not-fill-companion-extension-v4-7-4-now-supports-shadow-doms/p2)) while LastPass does not ([forum](https://community.logmein.com/t5/LastPass-Support-Discussions/Lastpass-is-not-compatible-with-ionic-4/m-p/238912/highlight/true#M12503)).

**Comment by ClaudioVR** [comment]
@wlee221 Regarding what you said: "you shouldn't need to update any code to use this workaround"...

does this also stand for Vue.js?

**Comment by wlee221** [comment]
@ClaudioVR, No let me clarify this a little bit.

- React
  - If you're using `withAuthenticator` HOC, no code change is needed.
  - If you're using `<AmplifyAuthenticator />` directly, wrap it with `AmplifyAuthContainer` like so: https://github.com/aws-amplify/amplify-js/issues/5782#issuecomment-828763711.
- Angular / Vue
  - You need to wrap around your `amplify-authenticator` component with `amplify-auth-container`:

```html
<amplify-auth-container>
  <amplify-authenticator></amplify-authenticator>
</amplify-auth-container>
```
I have a docs PR https://github.com/aws-amplify/docs/pull/3224 for this explanation, which I'll merge after our next JS release.

**Comment by fosrias** [comment]
> while we work on the rewrite.

@wlee221 Just to be crystal clear in this thread, is the re-write being done to accommodate a broad variety of password managers or not? Thanks.

**Comment by wlee221** [comment]
It will. The rewrite will stick with the best html practices to support a broad variety of password managers, which really boils down to not using shadow DOM.

**Comment by sammartinez** [comment]
https://github.com/aws-amplify/amplify-js/issues/8289

**Comment by tommulkins** [comment]
Just checking in to see if there's an update on this.  It appears the Amplify Admin UI suffers from the same problem.  At least AWS is dog-fooding their own product 😆

**Comment by ericclemmons** [comment]
@tommulkins It could be for a different reason! Can you open an issue here so they can improve that separately?
> https://github.com/aws-amplify/amplify-adminui

Thanks!

**Comment by thetrevdev** [comment]
Is there a rough ETA for a fix that would include Safari?

**Comment by patti-berchi** [comment]
@wlee221  I have tried the solution above and wrapped AmplifyAuthenticator  with AmplifyAuthContainer but I get the following error:

_Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: undefined. You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports. Check the render method of `AuthStateApp`._

I have installed the last unstable version (aws-amplify@4.1.2-unstable.12) and I'm importing AmplifyAuthenticator from "aws-amplify" + wrapping as follows
```
            <AmplifyAuthContainer>
              <AmplifyAuthenticator hideToast>
                <AmplifySignUp
                  slot="sign-up" formFields={[...]}  />
                <AmplifyConfirmSignUp
                  slot="confirm-sign-up" >
               </AmplifyConfirmSignUp>
              </AmplifyAuthenticator>
            </AmplifyAuthContainer>

```
Any ideas?

**Comment by wlee221** [comment]
@patti-berchi, I was able to get `AmplifyAuthContainer` working on latest React -- [codesandbox](https://codesandbox.io/s/serverless-smoke-woydr?file=/src/App.js) 

Do you still experience this issue? If so, can you open a new issue with more context? I'll be happy to take a look at it!

**Comment by patti-berchi** [comment]
@wlee221 Fixed! after hours of tweaking and trying different things just uninstalling and reinstalling the modules again seems to have solved it. Thanks though!
